### PR TITLE
FHAC-689 Implement Audit Query DAO calls for Audit Viewer

### PR DESCRIPTION
**Pull request by vdmehta06**
We can search audit entries using 3 below options:
1. By MessageId or RelatesTo
2. By EventOutcome or EventType or UserId or Remote HCID or Event Begin Date or Event End date
3. By AuditId
I cannot hook up this DAO query changes to Secured,Unsecured and Java Implementation classes as development is still not ready and it is addressed in upcoming other ticket and it is not addressed as a part of this PR
All the above 3 available search criteria are tested with dummy data and I have also created Junit methods but those are dummy values passed in test methods need to be updated based on database entries

### CODE REVIEW

**Comment by vdmehta06** [comment]
@alameluchidambaram @TabassumJafri

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 1]
```diff
@@ -161,4 +166,150 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+
+            LOG.info("Getting Record for Audit Viewer ");
+
+            // Build the criteria
+            Criteria queryCriteria = session.createCriteria(AuditRepositoryRecord.class);
+
+            if (NullChecker.isNotNullish(messageId) || NullChecker.isNotNullish(relatesTo)) {
+
+                if (NullChecker.isNotNullish(messageId) && (messageId.startsWith("urn:uuid:")
+                    || !messageId.startsWith("urn:uuid:"))) {
```
Comment (TabassumJafri) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: We have already check for messageId as NotNull in the above if  case. Do we need a second check on it?

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 2]
```diff
@@ -161,4 +166,150 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+                    if (messageId.endsWith("%")) {
+                        queryCriteria.add(Restrictions.like("messageId", messageId));
+                    } else {
+                        queryCriteria.add(Restrictions.like("messageId", messageId + "%"));
+                    }
+
+                }
+
+                if (NullChecker.isNotNullish(relatesTo)) {
+                    if (relatesTo.endsWith("%")) {
```
Comment (TabassumJafri) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Same comments as above

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 3]
```diff
@@ -161,4 +166,150 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+                        queryCriteria.add(Restrictions.like("relatesTo", relatesTo));
+
+                    } else {
+                        queryCriteria.add(Restrictions.like("relatesTo", relatesTo + "%"));
+
+                    }
+                }
+
+            } else if ((outcome != null && outcome >= 0) || NullChecker.isNotNullish(eventTypeList)
+                || NullChecker.isNotNullish(userId) || NullChecker.isNotNullish(remoteHcidList)
```
Comment (TabassumJafri) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: An else check is sufficient here. We can remove these checks in the else if block because all of the conditions are being checked in if blocks below.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 4]
```diff
@@ -161,4 +166,150 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+                }
+            }
+
+            queryCriteria.setProjection(Projections.projectionList()
+                .add(Projections.property("userId"))
+                .add(Projections.property("eventType"))
+                .add(Projections.property("eventId"))
+                .add(Projections.property("outcome"))
+                .add(Projections.property("eventTimeStamp"))
+                .add(Projections.property("remoteHcid"))
```
Comment (TabassumJafri) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: eventTimeStamp should be eventTimestamp. MySQL is case-insensitive for table and column name however it depends on the operating system as well. So your code probably works fine on windows but will have issues on Linux/unix machines.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 5]
```diff
@@ -161,4 +166,150 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+     * @param relatesTo
+     * @param outcome
+     * @param startDate
+     * @param userId
+     * @param eventTypeList
+     * @param remoteHcidList
+     * @param endDate
+     * @return List
+     */
+    public List queryAuditViewer(String messageId, String relatesTo, Integer outcome, List<String> eventTypeList,
```
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: We might want to remove the term "viewer" from here.  should just queryAuditRecords or something like that.  This class should have no idea of the usage, meaning the DAO call could be used by another service, not just a view.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 6]
```diff
@@ -161,4 +166,150 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+        try {
+            SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+            session = sessionFactory.openSession();
+
+            LOG.info("Getting Record for Audit Viewer ");
+
+            // Build the criteria
+            Criteria queryCriteria = session.createCriteria(AuditRepositoryRecord.class);
+
+            if (NullChecker.isNotNullish(messageId) || NullChecker.isNotNullish(relatesTo)) {
```
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: These if statements need to be refactored.  Should remove the top level check (relatesto or messageId don't equal null) and handle them separately.
Also, you have a check for if message id startwith or doesn't start with urn:uuid, that or will always be true.

**Comment by TabassumJafri** [comment]
Good with PR with no further comments.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 7]
```diff
@@ -161,4 +166,179 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+            if (NullChecker.isNotNullish(startDate) && NullChecker.isNotNullish(endDate)) {
+                SimpleDateFormat sdf = new SimpleDateFormat("MM/dd/yyyy");
+
+                queryCriteria.add(Restrictions.between("eventTimestamp", new Date(sdf.parse(startDate).getTime()),
+                    new Date(sdf.parse(endDate).getTime())));
+
+            } else if (NullChecker.isNotNullish(startDate) && NullChecker.isNullish(endDate)) {
+                SimpleDateFormat sdf = new SimpleDateFormat("MM/dd/yyyy");
+
+                queryCriteria.add(Restrictions.ge("eventTimestamp", new Date(sdf.parse(startDate).getTime())));
```
Comment (alameluchidambaram) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: what if the start date is not provided from UI?

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 8]
```diff
@@ -161,4 +166,184 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+            session = sessionFactory.openSession();
+
+            LOG.info("Getting Record for Audit Viewer ");
+
+            // Build the criteria
+            Criteria queryCriteria = session.createCriteria(AuditRepositoryRecord.class);
+
+            if (NullChecker.isNotNullish(messageId)) {
+                queryCriteria.add(Restrictions.eq("messageId", messageId));
+
```
Comment (alameluchidambaram) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Can you check prefix and if not prepend?

**Comment by alameluchidambaram** [comment]
I have reviewed this PR and good with merge.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 9]
```diff
@@ -161,4 +167,189 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+                queryCriteria.add(Restrictions.ge("eventTimestamp", new Date(sdf.parse(startDate).getTime())));
+
+            } else if (NullChecker.isNullish(startDate) && NullChecker.isNotNullish(endDate)) {
+                SimpleDateFormat sdf = new SimpleDateFormat("MM/dd/yyyy");
+
+                queryCriteria.add(Restrictions.le("eventTimestamp", new Date(sdf.parse(endDate).getTime())));
+
+            }
+
+            queryCriteria.setProjection(Projections.projectionList()
```
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: This set project is repeated in both methods and could be pulled out into a separate method.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 10]
```diff
@@ -161,4 +167,189 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+        return queryList;
+    }
+
+    /**
+     * This method does a query to database to get the Audit Log Blob Messages based on the auditId
+     *
+     * @param auditId
+     * @return List
+     */
+    public List queryAuditViewerByAuditId(String auditId) {
```
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: this method should just be called: queryByAuditId, need to remove the word "viewer".

Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Why is this method expecting a list?  If you provide an id, you should get one and only one result back.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 11]
```diff
@@ -161,4 +167,189 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+
+            // Build the criteria
+            Criteria queryCriteria = session.createCriteria(AuditRepositoryRecord.class);
+
+            if (NullChecker.isNotNullish(messageId)) {
+                if (messageId.startsWith(NhincConstants.WS_SOAP_HEADER_MESSAGE_ID_PREFIX)) {
+                    queryCriteria.add(Restrictions.eq("messageId", messageId));
+                } else {
+                    queryCriteria.add(Restrictions.eq("messageId", NhincConstants.WS_SOAP_HEADER_MESSAGE_ID_PREFIX + messageId));
+
```
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Alot of unnecessary spacing around here.  Can remove the empty lines such as line 196, 198, and 203

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 12]
```diff
@@ -161,4 +167,189 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+            SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+            session = sessionFactory.openSession();
+            LOG.info("Getting Record for Audit Viewer using auditId");
+
+            // Build the criteria
+            Criteria queryCriteria = session.createCriteria(AuditRepositoryRecord.class);
+            if (NullChecker.isNotNullish(auditId)) {
+                queryCriteria.add(Restrictions.eq("id", Integer.parseInt(auditId)));
+                queryCriteria.setProjection(Projections.property("message"));
+
```
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Remove empty line on 338.

File: Product/Production/Services/AuditRepositoryCore/src/test/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAOTest.java [hunk 13]
```diff
@@ -106,4 +120,63 @@ public void testQueryAuditRepositoryOnCriteria() {
 
     }
 
+    /**
+     * Test of queryAuditRepository Viewer with messageId and Outcome to test queryAuditViewer method, of class
+     * AuditRepositoryDAO .
+     */
+    @Test
+    public void testQueryAuditViewerByMessageIdAndCheckPrecedence() {
```
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/test/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAOTest.java]: Please remove the words "AuditViewer" from all of these test methods.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 14]
```diff
@@ -161,4 +167,189 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+            // if no criteria is passed then it will search full database with above mentioned columns in the result
+            queryList = queryCriteria.list();
+
+        } catch (HibernateException e) {
+            LOG.error("Exception in AuditLog.get() occurred due to :" + e.getLocalizedMessage(), e);
+        } finally {
+            // Actual contact insertion will happen at this step
+            if (session != null) {
+                session.flush();
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Why do we 'flush' in the finally, instead of 'commit' in the try?

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 15]
```diff
@@ -161,4 +167,189 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+     *
+     * @param outcome
+     * @param startDate
+     * @param userId
+     * @param eventTypeList
+     * @param remoteHcidList
+     * @param endDate
+     * @return List
+     */
+    public List queryByAuditValues(Integer outcome, List<String> eventTypeList, String userId, List<String> remoteHcidList, String startDate, String endDate) {
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Wrap this if you already have to make another commit; ignore otherwise.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 16]
```diff
@@ -161,4 +167,189 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+                .add(Projections.property("messageId")));
+
+            // if no criteria is passed then it will search full database with above mentioned columns in the result
+            queryList = queryCriteria.list();
+
+        } catch (HibernateException e) {
+            LOG.error("Exception in AuditLog.get() occurred due to :" + e.getLocalizedMessage(), e);
+        } finally {
+            // Actual contact insertion will happen at this step
+            if (session != null) {
```
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: The finally logic is repeated throughout, could be pulled into a helper method.

Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Per Jason's point, see the following class:

https://github.com/CONNECT-Solution/CONNECT/blob/CONNECT_integration/Product/Production/Services/DirectCore/src/main/java/gov/hhs/fha/nhinc/direct/messagemonitoring/dao/impl/MessageMonitoringDAOImpl.java#L224

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java (hunk 7 above)
Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: Remove the formatting and the dao should take in a date not a string.  The string to date conversion should happen at the service level above the dao.

File: Product/Production/Services/AuditRepositoryCore/src/test/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAOTest.java [hunk 17]
```diff
@@ -36,6 +39,9 @@
 import static org.junit.Assert.*;
 
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.logging.Level;
+import java.util.logging.Logger;
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/test/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAOTest.java]: I know you just didn't import JUL Logging, right? :stuck_out_tongue:

Comment (jsmith2) [inline Product/Production/Services/AuditRepositoryCore/src/test/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAOTest.java]: Good with PR, other than logging issue Chris brought up.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 18]
```diff
@@ -153,12 +157,178 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
+            }
+
+            if (NullChecker.isNotNullish(remoteHcidList)) {
+                queryCriteria.add(Restrictions.in("remoteHcid", remoteHcidList));
+            }
+
+            if (startDate != null && endDate != null) {
+                queryCriteria.add(Expression.between("eventTimestamp", new Date(startDate.getTime()),
+                    new Date(endDate.getTime())));
+            } else if (startDate != null && endDate == null) {
```
Comment (TabassumJafri) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java:250]: Do we need to create a new Date object here? startDate and endDate are already of type Date

**Comment by TabassumJafri** [comment]
Reviewed PR and good with merge

**Comment by christophermay07** [comment]
I've reviewed this PR, endorsing for merge without further comment.
